File: ios/RCTWebRTC/PalabraClient.m
#import "PalabraClient.h"
#import "WebRTCModule.h"
#import <AVFoundation/AVFoundation.h>

static const int kSampleRate = 16000;
static const int kChannels = 1;
static const int kChunkMs = 320;
static const int kChunkSamples = kSampleRate * kChunkMs / 1000;
static const int kChunkBytes = kChunkSamples * 2;

@interface PalabraClient () <NSURLSessionWebSocketDelegate, RTCAudioRenderer>

@property (nonatomic, strong) NSString *clientId;
@property (nonatomic, strong) NSString *clientSecret;
@property (nonatomic, strong) NSString *apiUrl;
@property (nonatomic, strong) NSString *sourceLang;
@property (nonatomic, strong) NSString *targetLang;

@property (nonatomic, strong) RTCAudioTrack *remoteTrack;

@property (nonatomic, strong) NSString *sessionId;
@property (nonatomic, strong) NSString *wsUrl;
@property (nonatomic, strong) NSString *publisherToken;

@property (nonatomic, strong) NSURLSession *urlSession;
@property (nonatomic, strong) NSURLSessionWebSocketTask *webSocket;

@property (nonatomic, strong) AVAudioEngine *audioEngine;
@property (nonatomic, strong) AVAudioPlayerNode *playerNode;
@property (nonatomic, strong) AVAudioFormat *audioFormat;

@property (nonatomic, assign) BOOL connected;
@property (nonatomic, assign) BOOL translating;
@property (nonatomic, assign) double originalVolume;

@property (nonatomic, strong) NSMutableData *audioBuffer;
@property (nonatomic, strong) NSLock *bufferLock;

@end

@implementation PalabraClient

- (instancetype)initWithClientId:(NSString *)clientId
                    clientSecret:(NSString *)clientSecret
                          apiUrl:(NSString *)apiUrl
                          module:(WebRTCModule *)module {
    self = [super init];
    if (self) {
        _clientId = clientId;
        _clientSecret = clientSecret;
        _apiUrl = apiUrl;
        _module = module;
        _connected = NO;
        _translating = NO;
        _originalVolume = 1.0;
        _audioBuffer = [NSMutableData new];
        _bufferLock = [NSLock new];
        
        NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
        _urlSession = [NSURLSession sessionWithConfiguration:config delegate:self delegateQueue:nil];
        
        [self setupAudioEngine];
    }
    return self;
}

- (BOOL)isConnected {
    return _connected;
}

- (BOOL)isTranslating {
    return _translating;
}

- (void)setupAudioEngine {
    self.audioEngine = [[AVAudioEngine alloc] init];
    self.playerNode = [[AVAudioPlayerNode alloc] init];
    [self.audioEngine attachNode:self.playerNode];
    
    self.audioFormat = [[AVAudioFormat alloc] initWithCommonFormat:AVAudioPCMFormatInt16
                                                        sampleRate:kSampleRate
                                                          channels:kChannels
                                                       interleaved:YES];
    
    [self.audioEngine connect:self.playerNode
                           to:self.audioEngine.mainMixerNode
                       format:self.audioFormat];
}

- (void)startWithTrack:(RTCAudioTrack *)remoteAudioTrack
            sourceLang:(NSString *)sourceLang
            targetLang:(NSString *)targetLang {
    
    if (self.translating) {
        return;
    }
    
    self.remoteTrack = remoteAudioTrack;
    self.sourceLang = sourceLang;
    self.targetLang = targetLang;
    
    self.originalVolume = remoteAudioTrack.source.volume;
    remoteAudioTrack.source.volume = 0;
    
    [self notifyConnectionState:@"connecting"];
    
    __weak typeof(self) weakSelf = self;
    [self createSessionWithCompletion:^(NSDictionary *session, NSError *error) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) return;
        
        if (error) {
            [strongSelf notifyError:error];
            strongSelf.remoteTrack.source.volume = strongSelf.originalVolume;
            return;
        }
        
        NSDictionary *data = session[@"data"];
        strongSelf.sessionId = data[@"id"];
        strongSelf.wsUrl = data[@"ws_url"];
        strongSelf.publisherToken = data[@"publisher"];
        
        NSLog(@"palabra_ws_url: %@", strongSelf.wsUrl);
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [strongSelf connectWebSocket];
        });
    }];
}

- (void)createSessionWithCompletion:(void (^)(NSDictionary *, NSError *))completion {
    NSString *urlStr = [NSString stringWithFormat:@"%@/session-storage/session", self.apiUrl];
    NSURL *url = [NSURL URLWithString:urlStr];
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"POST";
    [request setValue:self.clientId forHTTPHeaderField:@"ClientId"];
    [request setValue:self.clientSecret forHTTPHeaderField:@"ClientSecret"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    
    NSDictionary *body = @{
        @"data": @{
            @"subscriber_count": @0,
            @"publisher_can_subscribe": @YES
        }
    };
    
    NSError *jsonError;
    request.HTTPBody = [NSJSONSerialization dataWithJSONObject:body options:0 error:&jsonError];
    
    if (jsonError) {
        completion(nil, jsonError);
        return;
    }
    
    NSURLSessionDataTask *task = [[NSURLSession sharedSession] dataTaskWithRequest:request
        completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
            if (error) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    completion(nil, error);
                });
                return;
            }
            
            NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
            if (httpResponse.statusCode < 200 || httpResponse.statusCode >= 300) {
                NSError *httpError = [NSError errorWithDomain:@"PalabraClient"
                                                        code:httpResponse.statusCode
                                                    userInfo:@{NSLocalizedDescriptionKey: [NSString stringWithFormat:@"session_http_error_%ld", (long)httpResponse.statusCode]}];
                dispatch_async(dispatch_get_main_queue(), ^{
                    completion(nil, httpError);
                });
                return;
            }
            
            NSError *parseError;
            NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&parseError];
            
            dispatch_async(dispatch_get_main_queue(), ^{
                completion(json, parseError);
            });
        }];
    
    [task resume];
}

- (void)connectWebSocket {
    NSString *endpoint = [NSString stringWithFormat:@"%@?token=%@", self.wsUrl, self.publisherToken];
    NSLog(@"palabra_connecting_ws: %@", endpoint);
    
    NSURL *url = [NSURL URLWithString:endpoint];
    self.webSocket = [self.urlSession webSocketTaskWithURL:url];
    [self.webSocket resume];
    
    [self receiveMessage];
    
    self.connected = YES;
    self.translating = YES;
    
    [self.remoteTrack addRenderer:self];
    
    NSError *audioError;
    [self.audioEngine startAndReturnError:&audioError];
    if (!audioError) {
        [self.playerNode play];
    }
    
    [self notifyConnectionState:@"connected"];
    
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self sendSetTask];
    });
}

- (void)receiveMessage {
    if (!self.webSocket) return;
    
    __weak typeof(self) weakSelf = self;
    [self.webSocket receiveMessageWithCompletionHandler:^(NSURLSessionWebSocketMessage *message, NSError *error) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) return;
        
        if (error) {
            NSLog(@"palabra_ws_error: %@", error);
            dispatch_async(dispatch_get_main_queue(), ^{
                [strongSelf stop];
                [strongSelf notifyError:error];
            });
            return;
        }
        
        if (message.type == NSURLSessionWebSocketMessageTypeString) {
            [strongSelf handleMessage:message.string];
        }
        
        [strongSelf receiveMessage];
    }];
}

- (void)sendSetTask {
    if (!self.webSocket || !self.connected) return;
    
    NSDictionary *msg = @{
        @"message_type": @"set_task",
        @"data": @{
            @"input_stream": @{
                @"content_type": @"audio",
                @"source": @{
                    @"type": @"ws",
                    @"format": @"pcm_s16le",
                    @"sample_rate": @(kSampleRate),
                    @"channels": @(kChannels)
                }
            },
            @"output_stream": @{
                @"content_type": @"audio",
                @"target": @{
                    @"type": @"ws",
                    @"format": @"pcm_s16le"
                }
            },
            @"pipeline": @{
                @"transcription": @{
                    @"source_language": self.sourceLang
                },
                @"translations": @[@{
                    @"target_language": self.targetLang,
                    @"speech_generation": @{
                        @"voice_cloning": @NO
                    }
                }],
                @"allowed_message_types": @[
                    @"partial_transcription",
                    @"validated_transcription",
                    @"translated_transcription"
                ]
            }
        }
    };
    
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:msg options:0 error:&error];
    if (error) {
        NSLog(@"palabra_set_task_error: %@", error);
        return;
    }
    
    NSString *payload = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    NSLog(@"palabra_set_task: %@", payload);
    
    NSURLSessionWebSocketMessage *wsMsg = [[NSURLSessionWebSocketMessage alloc] initWithString:payload];
    [self.webSocket sendMessage:wsMsg completionHandler:^(NSError *error) {
        if (error) {
            NSLog(@"palabra_send_error: %@", error);
        }
    }];
}

- (void)handleMessage:(NSString *)text {
    NSError *error;
    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:[text dataUsingEncoding:NSUTF8StringEncoding]
                                                         options:0
                                                           error:&error];
    if (error) {
        NSLog(@"palabra_parse_error: %@", error);
        return;
    }
    
    NSString *type = json[@"message_type"] ?: @"";
    
    if ([type isEqualToString:@"output_audio_data"]) {
        NSDictionary *data = json[@"data"];
        NSString *audioBase64 = data[@"data"] ?: @"";
        if (audioBase64.length > 0) {
            NSData *audioData = [[NSData alloc] initWithBase64EncodedString:audioBase64 options:0];
            [self playAudio:audioData];
        }
    } else if ([type containsString:@"transcription"]) {
        NSDictionary *data = json[@"data"];
        NSDictionary *transcription = data[@"transcription"];
        if (transcription) {
            NSString *text = transcription[@"text"] ?: @"";
            NSString *lang = transcription[@"language"] ?: @"";
            BOOL isFinal = ![type isEqualToString:@"partial_transcription"];
            
            dispatch_async(dispatch_get_main_queue(), ^{
                [self notifyTranscription:text language:lang isFinal:isFinal];
            });
        }
    } else if ([type isEqualToString:@"error"]) {
        NSDictionary *data = json[@"data"];
        NSString *desc = data[@"desc"] ?: @"unknown";
        NSLog(@"palabra_error: %@", desc);
        NSError *err = [NSError errorWithDomain:@"PalabraClient" code:500 userInfo:@{NSLocalizedDescriptionKey: desc}];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self notifyError:err];
        });
    }
}

- (void)playAudio:(NSData *)audioData {
    if (!self.translating || !self.playerNode) return;
    
    AVAudioFrameCount frameCount = (AVAudioFrameCount)(audioData.length / 2);
    AVAudioPCMBuffer *buffer = [[AVAudioPCMBuffer alloc] initWithPCMFormat:self.audioFormat frameCapacity:frameCount];
    buffer.frameLength = frameCount;
    
    memcpy(buffer.int16ChannelData[0], audioData.bytes, audioData.length);
    
    [self.playerNode scheduleBuffer:buffer completionHandler:nil];
}

- (void)stop {
    if (!self.translating) return;
    
    self.translating = NO;
    self.connected = NO;
    
    if (self.remoteTrack) {
        [self.remoteTrack removeRenderer:self];
        self.remoteTrack.source.volume = self.originalVolume;
    }
    
    if (self.webSocket) {
        NSDictionary *endMsg = @{
            @"message_type": @"end_task",
            @"data": @{@"force": @NO}
        };
        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:endMsg options:0 error:nil];
        NSString *payload = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
        NSURLSessionWebSocketMessage *msg = [[NSURLSessionWebSocketMessage alloc] initWithString:payload];
        [self.webSocket sendMessage:msg completionHandler:nil];
        
        [self.webSocket cancelWithCloseCode:NSURLSessionWebSocketCloseCodeNormalClosure reason:nil];
        self.webSocket = nil;
    }
    
    [self.playerNode stop];
    [self.audioEngine stop];
    
    [self.bufferLock lock];
    [self.audioBuffer setLength:0];
    [self.bufferLock unlock];
    
    self.remoteTrack = nil;
    [self notifyConnectionState:@"disconnected"];
}

- (void)setTranslatedVolume:(double)volume {
    if (self.playerNode) {
        self.playerNode.volume = volume;
    }
}

- (void)sendAudioToPalabra:(AVAudioPCMBuffer *)buffer {
}

#pragma mark - RTCAudioRenderer

- (void)renderPCMBuffer:(AVAudioPCMBuffer *)buffer {
    if (!self.translating || !self.webSocket) return;
    
    NSData *resampled = [self resampleBuffer:buffer toRate:kSampleRate channels:kChannels];
    
    [self.bufferLock lock];
    [self.audioBuffer appendData:resampled];
    
    while (self.audioBuffer.length >= kChunkBytes) {
        NSData *chunk = [self.audioBuffer subdataWithRange:NSMakeRange(0, kChunkBytes)];
        NSMutableData *remaining = [NSMutableData dataWithData:[self.audioBuffer subdataWithRange:NSMakeRange(kChunkBytes, self.audioBuffer.length - kChunkBytes)]];
        self.audioBuffer = remaining;
        
        [self sendAudioChunk:chunk];
    }
    [self.bufferLock unlock];
}

- (NSData *)resampleBuffer:(AVAudioPCMBuffer *)buffer toRate:(int)dstRate channels:(int)dstChannels {
    int srcRate = (int)buffer.format.sampleRate;
    int srcChannels = (int)buffer.format.channelCount;
    int srcSamples = (int)buffer.frameLength;
    
    if (srcRate == dstRate && srcChannels == dstChannels) {
        if (buffer.format.commonFormat == AVAudioPCMFormatInt16) {
            return [NSData dataWithBytes:buffer.int16ChannelData[0] length:srcSamples * 2];
        }
    }
    
    float *srcData;
    if (buffer.format.commonFormat == AVAudioPCMFormatFloat32) {
        srcData = buffer.floatChannelData[0];
    } else {
        return [NSData data];
    }
    
    float *monoSrc = srcData;
    float *monoBuffer = NULL;
    if (srcChannels == 2 && buffer.floatChannelData[1]) {
        monoBuffer = malloc(srcSamples * sizeof(float));
        for (int i = 0; i < srcSamples; i++) {
            monoBuffer[i] = (buffer.floatChannelData[0][i] + buffer.floatChannelData[1][i]) / 2.0f;
        }
        monoSrc = monoBuffer;
    }
    
    int dstSamples = (int)((long)srcSamples * dstRate / srcRate);
    int16_t *dstData = malloc(dstSamples * sizeof(int16_t));
    
    for (int i = 0; i < dstSamples; i++) {
        float srcIdx = (float)i * (srcSamples - 1) / (dstSamples - 1);
        int idx0 = (int)srcIdx;
        int idx1 = MIN(idx0 + 1, srcSamples - 1);
        float frac = srcIdx - idx0;
        float sample = monoSrc[idx0] * (1 - frac) + monoSrc[idx1] * frac;
        dstData[i] = (int16_t)(sample * 32767.0f);
    }
    
    if (monoBuffer) free(monoBuffer);
    
    NSData *result = [NSData dataWithBytes:dstData length:dstSamples * 2];
    free(dstData);
    return result;
}

- (void)sendAudioChunk:(NSData *)chunk {
    if (!self.webSocket || !self.connected) return;
    
    NSString *base64 = [chunk base64EncodedStringWithOptions:0];
    NSDictionary *msg = @{
        @"message_type": @"input_audio_data",
        @"data": @{
            @"data": base64
        }
    };
    
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:msg options:0 error:nil];
    NSString *payload = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    
    NSURLSessionWebSocketMessage *wsMsg = [[NSURLSessionWebSocketMessage alloc] initWithString:payload];
    [self.webSocket sendMessage:wsMsg completionHandler:nil];
}

#pragma mark - Notifications

- (void)notifyConnectionState:(NSString *)state {
    if ([self.delegate respondsToSelector:@selector(palabraDidChangeConnectionState:)]) {
        [self.delegate palabraDidChangeConnectionState:state];
    }
}

- (void)notifyError:(NSError *)error {
    if ([self.delegate respondsToSelector:@selector(palabraDidFailWithError:)]) {
        [self.delegate palabraDidFailWithError:error];
    }
}

- (void)notifyTranscription:(NSString *)text language:(NSString *)lang isFinal:(BOOL)isFinal {
    if ([self.delegate respondsToSelector:@selector(palabraDidReceiveTranscription:)]) {
        [self.delegate palabraDidReceiveTranscription:@{
            @"text": text,
            @"language": lang,
            @"isFinal": @(isFinal)
        }];
    }
}

@end

File: android/src/main/java/com/oney/WebRTCModule/palabra/PalabraClient.java
package com.oney.WebRTCModule.palabra;

import android.content.Context;
import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioTrack;
import android.os.Handler;
import android.os.Looper;
import android.util.Base64;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.webrtc.AudioTrackSink;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;

public class PalabraClient implements AudioTrackSink {
    private static final String TAG = "PalabraClient";
    private static final int SAMPLE_RATE_IN = 16000;
    private static final int SAMPLE_RATE_OUT = 24000;
    private static final int CHANNELS = 1;
    private static final int CHUNK_MS = 320;
    private static final int CHUNK_SAMPLES = SAMPLE_RATE_IN * CHUNK_MS / 1000;
    private static final int CHUNK_BYTES = CHUNK_SAMPLES * 2;

    private final Context context;
    private final PalabraConfig config;
    private PalabraListener listener;

    private org.webrtc.AudioTrack remoteTrack;
    private OkHttpClient httpClient;
    private WebSocket webSocket;

    private AudioTrack audioPlayer;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    private String sessionId;
    private String wsUrl;
    private String publisherToken;

    private AtomicBoolean connected = new AtomicBoolean(false);
    private AtomicBoolean translating = new AtomicBoolean(false);

    private ByteArrayOutputStream audioBuffer = new ByteArrayOutputStream();
    private final Object bufferLock = new Object();

    public PalabraClient(Context context, PalabraConfig config) {
        this.context = context;
        this.config = config;
        this.httpClient = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build();
        setupAudioPlayer();
    }

    public void setListener(PalabraListener listener) {
        this.listener = listener;
    }

    public boolean isConnected() {
        return connected.get();
    }

    public boolean isTranslating() {
        return translating.get();
    }

    private void setupAudioPlayer() {
        int channelConfig = AudioFormat.CHANNEL_OUT_MONO;
        int audioFormat = AudioFormat.ENCODING_PCM_16BIT;
        int bufferSize = AudioTrack.getMinBufferSize(SAMPLE_RATE_OUT, channelConfig, audioFormat) * 2;

        audioPlayer = new AudioTrack(
            AudioManager.STREAM_VOICE_CALL,
            SAMPLE_RATE_OUT,
            channelConfig,
            audioFormat,
            bufferSize,
            AudioTrack.MODE_STREAM
        );
    }

    public void start(org.webrtc.AudioTrack remoteAudioTrack) {
        if (translating.get()) {
            return;
        }

        this.remoteTrack = remoteAudioTrack;
        remoteAudioTrack.setVolume(0);

        notifyConnectionState("connecting");

        executor.execute(() -> {
            try {
                JSONObject session = createSession();
                Log.d(TAG, "session_response: " + session.toString());
                JSONObject data = session.getJSONObject("data");
                sessionId = data.getString("id");
                wsUrl = data.getString("ws_url");
                publisherToken = data.getString("publisher");
                Log.d(TAG, "ws_url: " + wsUrl);

                mainHandler.post(this::connectWebSocket);
            } catch (Exception e) {
                Log.e(TAG, "session_create_failed", e);
                mainHandler.post(() -> {
                    if (remoteAudioTrack != null) {
                        remoteAudioTrack.setVolume(1.0);
                    }
                    notifyError(500, e.getMessage());
                });
            }
        });
    }

    private JSONObject createSession() throws IOException, JSONException {
        URL url = new URL(config.apiUrl + "/session-storage/session");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("ClientId", config.clientId);
        conn.setRequestProperty("ClientSecret", config.clientSecret);
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setDoOutput(true);

        JSONObject body = new JSONObject();
        JSONObject bodyData = new JSONObject();
        bodyData.put("subscriber_count", 0);
        bodyData.put("publisher_can_subscribe", true);
        body.put("data", bodyData);

        try (OutputStream os = conn.getOutputStream()) {
            os.write(body.toString().getBytes(StandardCharsets.UTF_8));
        }

        int responseCode = conn.getResponseCode();
        if (responseCode < 200 || responseCode >= 300) {
            throw new IOException("session_http_error_" + responseCode);
        }

        StringBuilder response = new StringBuilder();
        try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
            String line;
            while ((line = br.readLine()) != null) {
                response.append(line);
            }
        }

        return new JSONObject(response.toString());
    }

    private void connectWebSocket() {
        String endpoint = wsUrl + "?token=" + publisherToken;
        Log.d(TAG, "connecting_ws: " + endpoint);

        Request request = new Request.Builder()
            .url(endpoint)
            .build();

        webSocket = httpClient.newWebSocket(request, new WebSocketListener() {
            @Override
            public void onOpen(WebSocket ws, Response response) {
                Log.d(TAG, "ws_open");
                connected.set(true);
                translating.set(true);

                remoteTrack.addSink(PalabraClient.this);
                audioPlayer.play();

                mainHandler.post(() -> notifyConnectionState("connected"));

                mainHandler.postDelayed(() -> sendSetTask(), 500);
            }

            @Override
            public void onMessage(WebSocket ws, String text) {
                handleMessage(text);
            }

            @Override
            public void onFailure(WebSocket ws, Throwable t, Response response) {
                Log.e(TAG, "ws_error", t);
                mainHandler.post(() -> {
                    stop();
                    notifyError(500, t.getMessage());
                });
            }

            @Override
            public void onClosed(WebSocket ws, int code, String reason) {
                Log.d(TAG, "ws_closed: " + code);
                mainHandler.post(() -> stop());
            }
        });
    }

    private void sendSetTask() {
        if (webSocket == null || !connected.get()) {
            return;
        }

        try {
            JSONObject msg = new JSONObject();
            msg.put("message_type", "set_task");

            JSONObject data = new JSONObject();

            JSONObject inputStream = new JSONObject();
            inputStream.put("content_type", "audio");
            JSONObject source = new JSONObject();
            source.put("type", "ws");
            source.put("format", "pcm_s16le");
            source.put("sample_rate", SAMPLE_RATE_IN);
            source.put("channels", CHANNELS);
            inputStream.put("source", source);
            data.put("input_stream", inputStream);

            JSONObject outputStream = new JSONObject();
            outputStream.put("content_type", "audio");
            JSONObject target = new JSONObject();
            target.put("type", "ws");
            target.put("format", "pcm_s16le");
            outputStream.put("target", target);
            data.put("output_stream", outputStream);

            JSONObject pipeline = new JSONObject();

            JSONObject transcription = new JSONObject();
            transcription.put("source_language", config.sourceLang);
            pipeline.put("transcription", transcription);

            JSONArray translations = new JSONArray();
            JSONObject translation = new JSONObject();
            translation.put("target_language", config.targetLang);
            JSONObject speechGen = new JSONObject();
            speechGen.put("voice_cloning", false);
            translation.put("speech_generation", speechGen);
            translations.put(translation);
            pipeline.put("translations", translations);

            JSONArray allowedTypes = new JSONArray();
            allowedTypes.put("partial_transcription");
            allowedTypes.put("validated_transcription");
            allowedTypes.put("translated_transcription");
            pipeline.put("allowed_message_types", allowedTypes);

            data.put("pipeline", pipeline);
            msg.put("data", data);

            String payload = msg.toString();
            Log.d(TAG, "set_task: " + payload);
            webSocket.send(payload);
        } catch (JSONException e) {
            Log.e(TAG, "set_task_error", e);
        }
    }

    private void handleMessage(String text) {
        try {
            JSONObject json = new JSONObject(text);
            String type = json.optString("message_type", "");

            if ("output_audio_data".equals(type)) {
                JSONObject data = json.getJSONObject("data");
                String audioBase64 = data.optString("data", "");
                if (!audioBase64.isEmpty()) {
                    byte[] audioBytes = Base64.decode(audioBase64, Base64.DEFAULT);
                    if (audioPlayer != null && translating.get()) {
                        audioPlayer.write(audioBytes, 0, audioBytes.length);
                    }
                }
            } else if (type.contains("transcription")) {
                JSONObject data = json.getJSONObject("data");
                JSONObject transcription = data.optJSONObject("transcription");
                if (transcription != null) {
                    String txt = transcription.optString("text", "");
                    String lang = transcription.optString("language", "");
                    boolean isFinal = !"partial_transcription".equals(type);
                    mainHandler.post(() -> notifyTranscription(txt, lang, isFinal));
                }
            } else if ("error".equals(type)) {
                JSONObject data = json.optJSONObject("data");
                String desc = data != null ? data.optString("desc", "unknown") : "unknown";
                Log.e(TAG, "palabra_error: " + desc);
                mainHandler.post(() -> notifyError(500, desc));
            }
        } catch (JSONException e) {
            Log.e(TAG, "msg_parse_error", e);
        }
    }

    public void stop() {
        if (!translating.getAndSet(false)) {
            return;
        }

        connected.set(false);

        if (remoteTrack != null) {
            try {
                remoteTrack.removeSink(this);
                remoteTrack.setVolume(1.0);
            } catch (Exception e) {
                Log.w(TAG, "stop_track_cleanup_error: " + e.getMessage());
            }
        }

        if (webSocket != null) {
            try {
                JSONObject endMsg = new JSONObject();
                endMsg.put("message_type", "end_task");
                endMsg.put("data", new JSONObject().put("force", false));
                webSocket.send(endMsg.toString());
            } catch (JSONException e) {
                Log.e(TAG, "end_task_error", e);
            }
            try {
                webSocket.close(1000, "stop");
            } catch (Exception e) {
                Log.w(TAG, "websocket_close_error: " + e.getMessage());
            }
            webSocket = null;
        }

        if (audioPlayer != null) {
            try {
                audioPlayer.stop();
            } catch (Exception e) {
                Log.w(TAG, "audio_player_stop_error: " + e.getMessage());
            }
        }

        synchronized (bufferLock) {
            audioBuffer.reset();
        }

        remoteTrack = null;
        notifyConnectionState("disconnected");
    }

    @Override
    public void onData(ByteBuffer audioData, int bitsPerSample, int sampleRate, int channels, int frames, long timestamp) {
        if (!translating.get() || webSocket == null) {
            return;
        }

        byte[] samples = new byte[audioData.remaining()];
        audioData.get(samples);

        byte[] resampled = resample(samples, sampleRate, channels, SAMPLE_RATE_IN, CHANNELS);

        synchronized (bufferLock) {
            try {
                audioBuffer.write(resampled);

                while (audioBuffer.size() >= CHUNK_BYTES) {
                    byte[] chunk = new byte[CHUNK_BYTES];
                    byte[] all = audioBuffer.toByteArray();
                    System.arraycopy(all, 0, chunk, 0, CHUNK_BYTES);

                    audioBuffer.reset();
                    if (all.length > CHUNK_BYTES) {
                        audioBuffer.write(all, CHUNK_BYTES, all.length - CHUNK_BYTES);
                    }

                    sendAudioChunk(chunk);
                }
            } catch (IOException e) {
                Log.e(TAG, "buffer_error", e);
            }
        }
    }

    private byte[] resample(byte[] input, int srcRate, int srcChannels, int dstRate, int dstChannels) {
        if (srcRate == dstRate && srcChannels == dstChannels) {
            return input;
        }

        int srcSamples = input.length / (2 * srcChannels);
        int dstSamples = (int) ((long) srcSamples * dstRate / srcRate);

        short[] srcData = new short[srcSamples * srcChannels];
        ByteBuffer.wrap(input).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer().get(srcData);

        short[] monoSrc = srcData;
        if (srcChannels == 2) {
            monoSrc = new short[srcSamples];
            for (int i = 0; i < srcSamples; i++) {
                monoSrc[i] = (short) ((srcData[i * 2] + srcData[i * 2 + 1]) / 2);
            }
        }

        short[] dstData = new short[dstSamples];
        for (int i = 0; i < dstSamples; i++) {
            float srcIdx = (float) i * (monoSrc.length - 1) / (dstSamples - 1);
            int idx0 = (int) srcIdx;
            int idx1 = Math.min(idx0 + 1, monoSrc.length - 1);
            float frac = srcIdx - idx0;
            dstData[i] = (short) (monoSrc[idx0] * (1 - frac) + monoSrc[idx1] * frac);
        }

        byte[] output = new byte[dstSamples * 2];
        ByteBuffer.wrap(output).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer().put(dstData);
        return output;
    }

    private void sendAudioChunk(byte[] chunk) {
        if (webSocket == null || !connected.get()) {
            return;
        }

        try {
            JSONObject msg = new JSONObject();
            msg.put("message_type", "input_audio_data");
            JSONObject data = new JSONObject();
            data.put("data", Base64.encodeToString(chunk, Base64.NO_WRAP));
            msg.put("data", data);
            webSocket.send(msg.toString());
        } catch (JSONException e) {
            Log.e(TAG, "send_audio_error", e);
        }
    }

    private void notifyConnectionState(String state) {
        if (listener != null) {
            listener.onConnectionState(state);
        }
    }

    private void notifyError(int code, String message) {
        if (listener != null) {
            listener.onError(code, message);
        }
    }

    private void notifyTranscription(String text, String lang, boolean isFinal) {
        if (listener != null) {
            listener.onTranscription(text, lang, isFinal);
        }
    }
}
